mesonbuild/backend/backends.py

```diff
@@ -607,29 +607,13 @@ def create_msvc_pch_implementation(self, target, lang, pch_header):
 
     @staticmethod
     def escape_extra_args(compiler, args):
-        # No extra escaping/quoting needed when not running on Windows
-        if not mesonlib.is_windows():
-            return args
+        # all backslashes in defines are doubly-escaped
         extra_args = []
-        # Compiler-specific escaping is needed for -D args but not for any others
-        if isinstance(compiler, VisualStudioLikeCompiler):
-            # MSVC needs escaping when a -D argument ends in \ or \"
-            for arg in args:
-                if arg.startswith('-D') or arg.startswith('/D'):
-                    # Without extra escaping for these two, the next character
-                    # gets eaten
-                    if arg.endswith('\\'):
-                        arg += '\\'
-                    elif arg.endswith('\\"'):
-                        arg = arg[:-2] + '\\\\"'
-                extra_args.append(arg)
-        else:
-            # MinGW GCC needs all backslashes in defines to be doubly-escaped
-            # FIXME: Not sure about Cygwin or Clang
-            for arg in args:
-                if arg.startswith('-D') or arg.startswith('/D'):
-                    arg = arg.replace('\\', '\\\\')
-                extra_args.append(arg)
+        for arg in args:
+            if arg.startswith('-D') or arg.startswith('/D'):
```

Comment:
Patch commentary says we are doing this to all arguments, but I think maybe I was just confused.

Reply:
Fixed patch comment.

Reply:
I'm not sure exactly how Meson output and Ninja processing results in a CreateProcess call on Windows but I suspect that the issue is with general quoting on not specifically the -D option although it is more likely to cause problems.

The Windows CreateProcess requires a single command line string of combined arguments. The rules are that backslash is not escaped unless followed by a double quote. Double quote always need escaping by prefixing a backslash. This means doubling the number of backslashes and adding one more if there is a sequence of backslashes followed by a double quote, or just inserting a backslash if there is a single double quote.

Also note that a pair of double quotes collapse into a single double quote, but with the above rules each of these would be prefixed by a backslash thus avoiding that collapse. The pair would be treated differently prior to 2008.

http://daviddeley.com/autohotkey/parameters/parameters.htm

Reply:
Ninja probably already do this for arguments it believes to be files.

Reply:
> I suspect that the issue is with general quoting on not specifically the -D option although it is more likely to cause problems.

This suspicion is correct.

> The rules are that backslash is not escaped unless followed by a double quote. Double quote always need escaping by prefixing a backslash. This means doubling the number of backslashes and adding one more if there is a sequence of backslashes followed by a double quote, or just inserting a backslash if there is a single double quote.

This is now implemented in cmd_quote() in ninjabackend.py.  We didn't need it before this PR, because we always used response files.

I believe if you look at the git history of this function, you'll see that it's intended to make things like the following work as intended:

```
c_args: '-DPATH=C:\now\thats\a\path'
```

unfortunately, this means that the following doesn't work as intended:

```
c_args: '-DCHAR=\n'
```

I think additions to the meson language would be needed to allow both to be expressed unambigiously.